### PR TITLE
Assemble depex

### PR DESCRIPTION
**Pull request by GanShun**

### CODE REVIEW

File: pkg/uefi/section.go [hunk 1]
```diff
@@ -170,6 +170,16 @@ var DepExOpCodes = map[byte]DepExOpCode{
 	0x9: "SOR",
 }
 
+// DepExNamesToOpCodes maps the operation back to the code.
+var DepExNamesToOpCodes map[DepExOpCode]byte
+
+func init() {
+	DepExNamesToOpCodes = make(map[DepExOpCode]byte)
```
Comment (rjoleary) [inline pkg/uefi/section.go]: Delete this line

File: pkg/uefi/section.go [hunk 2]
```diff
@@ -170,6 +170,16 @@ var DepExOpCodes = map[byte]DepExOpCode{
 	0x9: "SOR",
 }
 
+// DepExNamesToOpCodes maps the operation back to the code.
+var DepExNamesToOpCodes map[DepExOpCode]byte
```
Comment (rjoleary) [inline pkg/uefi/section.go]:
```suggestion
var DepExNamesToOpCodes = map[DepExOpCode]byte{}
```

File: pkg/visitors/assemble_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 the LinuxBoot Authors. All rights reserved
```
Comment (rjoleary) [inline pkg/visitors/assemble_test.go]: 2019

File: pkg/visitors/assemble.go [hunk 4]
```diff
@@ -242,6 +242,35 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
...
+					opcode, ok := uefi.DepExNamesToOpCodes[d.OpCode]
+					if !ok {
+						return fmt.Errorf("unable to map depex opcode string to opcode, string was: %v",
+							d.OpCode)
+					}
+					newBuf = append(newBuf, opcode)
+					// Sanity checks.
+					switch d.OpCode {
+					case "BEFORE", "AFTER", "PUSH":
+						// GUID should not be nil.
```
Comment (rjoleary) [inline pkg/visitors/assemble.go]:
```suggestion
						// GUID must not be nil.
```

File: pkg/visitors/assemble.go [hunk 5]
```diff
@@ -242,6 +242,35 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
...
+						if d.GUID == nil {
+							return fmt.Errorf("depex opcode %v should not have nil guid",
+								d.OpCode)
+						}
+						newBuf = append(newBuf, d.GUID[:]...)
+					default:
+						// GUID should be nil.
```
Comment (rjoleary) [inline pkg/visitors/assemble.go]:
```suggestion
						// GUID must be nil.
```

File: pkg/visitors/assemble.go [hunk 6]
```diff
@@ -242,6 +242,35 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
...
+						return fmt.Errorf("unable to map depex opcode string to opcode, string was: %v",
+							d.OpCode)
+					}
+					newBuf = append(newBuf, opcode)
+					// Sanity checks.
+					switch d.OpCode {
+					case "BEFORE", "AFTER", "PUSH":
+						// GUID should not be nil.
+						if d.GUID == nil {
+							return fmt.Errorf("depex opcode %v should not have nil guid",
```
Comment (rjoleary) [inline pkg/visitors/assemble.go]:
```suggestion
							return fmt.Errorf("depex opcode %v must not have nil guid",
```

File: pkg/visitors/assemble.go [hunk 7]
```diff
@@ -242,6 +242,35 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
...
+								d.OpCode)
+						}
+						newBuf = append(newBuf, d.GUID[:]...)
+					default:
+						// GUID should be nil.
+						if d.GUID != nil {
+							return fmt.Errorf("depex opcode %v should not have a guid! got %v",
```
Comment (rjoleary) [inline pkg/visitors/assemble.go]:
```suggestion
							return fmt.Errorf("depex opcode %v must not have a guid! got %v",
```

File: pkg/visitors/assemble.go [hunk 8]
```diff
@@ -242,6 +242,35 @@ func (v *Assemble) Visit(f uefi.Firmware) error {
 				binary.LittleEndian.PutUint16(newBuf, f.BuildNumber)
 				newBuf = append(newBuf, unicode.UTF8ToUCS2(f.Version)...)
 				f.SetBuf(newBuf)
+			case uefi.SectionTypeDXEDepEx, uefi.SectionTypePEIDepEx,
+				uefi.SectionMMDepEx:
```
Comment (rjoleary) [inline pkg/visitors/assemble.go]: There are a few more sanity checks you could do, but I'm not sure if they are worth it. For example, PEIDepEx does not support all the opcodes. Also the PI spec says that "NoteThe PEI Foundation will only support the evaluation of dependency expressions that are less than or equal to 256 terms"

Comment (GanShun) [reply pkg/visitors/assemble.go]: I'll add a TODO

**Pull request merged**